### PR TITLE
Containerd imports handled by k0s

### PR DESCRIPTION
**Pull request by jnummelin**
## Description

As noted in https://github.com/k0sproject/k0s/pull/2604#issuecomment-1400173518 the current way containerd handles imports for CRI configs produces less than expected results.

In this PR I propose an alternative where k0s handles the imports:
- k0s reads all "partials" from `/etc/k0s/containerd.d/*.toml`
- if a partial has CRI configs k0s will append them in memory with the default CRI config (using runc)
- if a partial does NOT have CRI config k0s will add it as normal `import` in containerd config
- once all CRI configs are appended in memory, k0s will drop the cri configs into `/run/k0s/containerd-cri.toml` file and add that as normal import in main containerd config toml

As noted in our proposed fix in upstream [containerd PR](https://github.com/containerd/containerd/pull/8056), it's not going to be included in 1.x releases. 😞  Hence the need to handle this at k0s side.

### CODE REVIEW

**Comment by alegrey91** [comment]
Hi @jnummelin,
IMHO the idea is fine and works properly, but it would not be preferable to try to fix this in **containerd**? In this way, you could have a better separation of duties. WDYT? Am I missing something?

**Comment by jnummelin** [comment]
> IMHO the idea is fine and works properly, but it would not be preferable to try to fix this in containerd?

Yes, of course. This PR is kinda like plan-B or an intermediary solution in case we cannot fix containerD side of things. Or if our proposal will not get traction in containerD. @juanluisvaladas is already working on the fix on containerD side.

File: pkg/component/worker/containerd.go [hunk 1]
```diff
@@ -91,14 +100,43 @@ func (c *ContainerD) Start(_ context.Context) error {
 
 func (c *ContainerD) setupConfig() error {
 	// If the config file exists, use it as-is
-	if file.Exists(confPath) {
-		return nil
-	}
+	// FIXME We need to figure out how to overwrite this only if it's fully k0s managed
```
Comment (makhov) [inline pkg/component/worker/containerd.go:103]: Do we need it? I think everything in the `/etc/k0s` directory is k0s managed

File: pkg/component/worker/containerd/runtime.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            IoGid = 0
+            IoUid = 0
+            NoNewKeyring = false
+            NoPivotRoot = false
+            Root = ""
+            ShimCgroup = ""
+            SystemdCgroup = false
+`
+
+const importsPath = "/etc/k0s/containerd.d/*.toml"
```
Comment (makhov) [inline pkg/component/worker/containerd/runtime.go:98]: We use the same path here and in the `containerd.go`. Probably will need to refactor this and pass the path as an argument or config value.

File: pkg/component/worker/containerd/runtime.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+			c.log.Debugf("found CRI plugin config in %s, appending to CRI config", file)
+			// Append all CRI plugin configs into single file
+			// and add it to imports
+			criConfigBuffer.WriteString(fmt.Sprintf("# appended from %s\n", file))
+			_, err := criConfigBuffer.Write(data)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			c.log.Debugf("adding %s as-is to imports", file)
```
Comment (makhov) [inline pkg/component/worker/containerd/runtime.go:150]: I think it's ok for now, but in the future, we may run into the same issue with another section.

File: pkg/component/worker/containerd/runtime.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			// Add file to imports
+			imports = append(imports, file)
+		}
+	}
+
+	// Write the CRI config to a file and add it to imports
+	err = os.WriteFile(containerdCRIConfigPath, criConfigBuffer.Bytes(), 0644)
```
Comment (makhov) [inline pkg/component/worker/containerd/runtime.go:157]: Looks a bit fishy to me. By the name of the function, I wouldn't expect a file writing here. Not sure, how to improve it tho, so we can leave it as is for now.

**Review by makhov — COMMENTED** [review]
LGTM in general, left a few comments

**Comment by jnummelin** [comment]
Upstream containerd is not willing to accept our proposed fix into 1.x series and is targeting the fix for only 2.0

Hence I'll start to work on this to make it more robust and prettify the code a bit. 😄

File: pkg/component/worker/containerd.go [hunk 5]
```diff
@@ -91,14 +100,43 @@ func (c *ContainerD) Start(_ context.Context) error {
...
+			Imports []string
+		}{
+			Imports: imports,
+		},
+	}
+	if err := tw.WriteToBuffer(output); err != nil {
+		return fmt.Errorf("can't create containerd config: %v", err)
+	}
+	if err := file.WriteContentAtomically(confPath, output.Bytes(), 0644); err != nil {
+		return err
```
Comment (mikhail-sakhnov) [inline pkg/component/worker/containerd.go:135]: I'd suggest use wrapping (preferable) or return as is without if.

File: pkg/component/worker/containerd.go (hunk 1 above)
Comment (jnummelin) [reply pkg/component/worker/containerd.go]: Yes, we need the logic to handle this. All previous versions (before this PR) will treat any existing `/etc/k0s/containerd.toml` file as user managed and we should NOT overwrite those by default.

File: pkg/component/worker/containerd/runtime.go (hunk 3 above)
Comment (jnummelin) [reply pkg/component/worker/containerd/runtime.go]: > we may run into the same issue with another section

Not sure I get what you mean by this? Do you mean some other section than CRI configs? We _may_ run into those cases for sure, but untill we know how those sections work, or are expected to work, I would not put any logic in place.

**Comment by mikhail-sakhnov** [comment]
looks reasonably good, though there are couple of linting issues and non-addressed FIXME comment

File: pkg/component/worker/containerd.go [hunk 6]
```diff
@@ -90,18 +108,109 @@ func (c *ContainerD) Start(_ context.Context) error {
 }
 
 func (c *ContainerD) setupConfig() error {
-	// If the config file exists, use it as-is
-	if file.Exists(confPath) {
+	// Check if the config file is user managed
+	// If it is, we should not touch it
+	k0sManaged, err := isK0sManagedConfig(confPath)
+	if err != nil {
+		return err
```
Comment (mikhail-sakhnov) [inline pkg/component/worker/containerd.go:115]: I suggest either wrapping errors here or on the caller level (#L89)

File: pkg/component/worker/containerd.go [hunk 7]
```diff
@@ -90,18 +108,109 @@ func (c *ContainerD) Start(_ context.Context) error {
...
+		if strings.Contains(scanner.Text(), magicMarker) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func isPre1_27ManagedConfig(path string) (bool, error) {
+	// Check MD5 sum of the config file
+	// If it matches the pre 1.26 config, it's k0s managed
```
Comment (mikhail-sakhnov) [inline pkg/component/worker/containerd.go:192]: either comment or variable name should be changed (pre 1.26, pre1_27)

**Review by mikhail-sakhnov — CHANGES_REQUESTED** [review]

**Comment by jnummelin** [comment]
Marked as a draft till I sort out how to document this

**Review by juanluisvaladas — APPROVED** [review]
LGTM.

Docs are clear and I think there haven't been more recent changes.

**Comment by jnummelin** [comment]
@juanluisvaladas Sorry, need a re-review as I had stale (== non pushed) commit for the dynamic re-loading of containerd in my local branch... 🤦

File: pkg/component/worker/containerd.go [hunk 8]
```diff
@@ -148,6 +158,69 @@ func (c *ContainerD) setupConfig() error {
...
+	log := logrus.WithFields(logrus.Fields{"component": "containerd", "phase": "restart"})
+
+	log.Info("restart requested")
+	if err := c.setupConfig(); err != nil {
+		log.WithError(err).Warn("failed to resolve config")
+		return
+	}
+
+	p := c.supervisor.GetProcess()
+	if err := p.Signal(syscall.SIGHUP); err != nil {
```
Comment (juanluisvaladas) [inline pkg/component/worker/containerd.go:219]: By sending SIGHUP we're killing the process rather than reloading it, so I would send SIGTERM but it's a nitpick. If you agree and want to change it ping me on slack after the change.

**Review by juanluisvaladas — APPROVED** [review]
The nitpick is really a nit so LGTM

File: pkg/component/worker/containerd.go (hunk 8 above)
Comment (jnummelin) [reply pkg/component/worker/containerd.go:219]: I guess my rationale for `SIGHUP` was that it is the "standard" for reloading. In containerd there's no realoading support per se, the only way is to restart the process. But having `SIGHUP` do that kinda makes sense, especially if containerd ever implement actual reload using `SIGHUP`. Based on the comments in https://github.com/containerd/containerd/issues/4478, it seems there's some "plans" to actually implement reloading.

TL;DR; I'd keep it as `SIGHUP` for "forward compatibility"

Comment (juanluisvaladas) [reply pkg/component/worker/containerd.go:219]: OK that's fair enough.

**Comment by jnummelin** [comment]
@mikhail-sakhnov Could you re-review, cannot merge as you had requested changes (which are addressed now)

**Review by mikhail-sakhnov — APPROVED** [review]
